### DISCUSSION TITLE
Python 타입 힌트

### DISCUSSION THREAD

**Discussion by Kirade** [8percent/styleguide#18] [General]
[PEP484](https://www.python.org/dev/peps/pep-0484/)

저희는 현재시점 까지는 파이썬의 타입 힌트를 적극적으로 활용하고 있지 않습니다.
타입 힌트가 필요한 이유에 대해 논의해보고 싶어요.

**Comment by donghyun0218** [comment]
제가 타입힌트를 사용하는 가장 큰 이유는 두 가지가 있습니다.
* 코드 이해를 위한 시간 단축
  * 특정 모듈(클래스, 메서드)을 사용할 때 전달해야 하는 인자 또는 반환 값에 대한 타입을 알기 위해 내부 코드를 읽어야 하는 상황이 종종 있습니다. 타입힌트를 사용하면 이런 상황에 시간을 절약할 수 있습니다. 
  * 물론 독스트링으로도 같은 역할을 할 수 있지만, 독스트링은 사람마다 작성방식이 미묘하게 다른 부분이 있을 수 있기 때문에 타입힌트가 형식 관리에 용이한 것 같습니다.
* IDE 호환
  * 저희가 사용하는 Pycharm 에서 타입힌트를 사용할 때, 개발 단계에서 일어날 수 있는 실수들을 빠르게 발견할 수 있습니다.

**Reply by Kirade** [reply]
저희 함수에 docstring 에 입력하는 매개변수 타입을 대체하는것도 고려해볼 수 있겠군요 👍

**Comment by wer153** [comment]
일반론적으로 아래와 같은 장단을 가지고 있다고 비춰집니다.

장점
- 에러방지
- 코드자체로서 문서화 기능
- IDE 활용성에 도움
- 덕타이핑 부작용 최소화(디버깅에 유리)

단점
- 추가 비용
- 모던 파이썬 요구(3.0이상, 해당없음)
- 초기 생산성 감소

> Type hints help document your code. Traditionally, you would use docstrings if you wanted to document the expected types of a function’s arguments. This works, but as there is no standard for docstrings (despite PEP 257), they can’t be easily used for automatic checks.
[link](https://realpython.com/lessons/pros-and-cons-type-hints/)


추가로 소요되는 비용이 낮고 효용이 높은 곳에 타입힌팅이 적용되는게  좋다고 생각합니다. 
명백하게 효용이 높을 때는 코드의 명확성을 높여줘서 가독성을 높여줄 경우입니다.
- 코드의 타입을 특정하기 어려운 변수명 보완
- 함수명이 담지 못하는 불분명한 반환타입 보완
- 여러 곳에서 쓰이는 모듈에 안정성 부여
- 고장나면 안되는 핵심 모듈

**Reply by Kirade** [reply]
설명 감사합니다.
시스템이 커지게 될 수록 관리하기 어려워지는 코드들의 안정성을 높이는데 도움이 될 수 있겠네요.

**Comment by kyungjunleeme** [comment]
저는 간단하게 view에서만 써보고, mypy는 안써봤지만, 관심은 많습니다.

- https://xo.dev/python-3-type-annotation-and-typing/
- https://www.youtube.com/watch?v=KWjGj_0lPro&t=592s
- https://blog.lablup.com/posts/2021/11/12/python-3.10-overview 

- 활용한다면 pre-commit 시에 점진적으로(앱별로 실행 후 확대) mypy를 실행해보는것도 어떨까 싶네요. 

유사한 맥락으로 typescript가 bug 발생률을 낮춰준다는 논문도 있습니다.
[typestudy.pdf](https://github.com/8percent/styleguide/files/7782525/typestudy.pdf)

- 핑퐁팀의 파이썬문화
> https://blog.pingpong.us/python-in-pingpong/

**Reply by Kirade** [reply]
mypy 를 사용하게 된다면, 현재 travis 에서 lint 체크를 하는 워커가 하나 있는데, 같이 해줄 수 있을듯해요.

**Reply by kyungjunleeme** [reply]
@Kirade 넵. 
.travis.yml 내에서 flake 사용하신 부분을 말씀하시는 것 같은데, travis의 .yml 문법 공부도 염두해놓겠습니다.

**Comment by Kirade** [comment]
타입 힌트를 사용하는데 있어, 사용에 명확한 기준이 필요 할 것 같다고 처음에 생각했지만
여러 여러 장단점을 고려해보았을 때, 가볍게 시작해도 좋을것 같다는 생각이 들었습니다.

기존 코드의 대부분은 타입 힌트가 존재하지 않는 상태이기 때문에, 타입 힌트 사용에 가이드가 존재하지는 않는 상태입니다.
따라서 코드 작성자가 타입 힌트가 필요하다고 생각되는 지점에 사용해 보고, 데이터를 쌓아가는 방향도 괜찮을 것 같습니다.

이후 타입 힌트가 어느정도 활용되면 다시 한 번 저희들의 타입 힌트 방식 사용 방식에 대해 논의해보기로해요.

**Comment by cbscsm** [comment]
타입힌트는 생산성 향상에 도움이 된다는 견해도 있습니다(TypeScript가 대표적).
미리 잠재 오류를 잡아내어 디버깅 시간을 줄여준다는 것 외에도 코드 작성 시 다음과 같은 이점이 있습니다.
1. IDE에서 보다 정확한 자동완성을 진행할 수 있게 해줌,
1. 호출 계층을 깊게 따라가지 않고도 타입을 확신할 수 있음
1. 일부 라이브러리(Pydantic 등)에서는 타입힌트를 기반으로 값을 validation하거나 문서를 작성해주기도 함

저는 3.6 때부터 타입힌트를 쓰기 시작했는데,
개인 적인 작업에도 항상 타입 힌트를 쓸 정도로 만족하고 있습니다.

특히 Python 3.10 부터 보다 더 직관적이고 편리한 타입 어노테이션이 가능해졌으므로(
- PEP604: union 연산자 지원
- PEP612: 파라미터 스펙 변수
- PEP613: 명시적 타입 별칭
- PEP647: 사용자 정의 타입가드
- PEP563(지연된 어노테이션 평가)이 기본값이 됨(Python 3.7에서 등장한 이후 3.9까지는 future를 통해 적용)
- Python 3.9에서 PEP585: 표준 콜렉션의 제너릭 지원(클래스 레벨 `__getitem__`)

), 신규 혹은 수정되는 코드에 손 가는 곳 부터, 강제하지는 않고 자율적으로 적용해 보는 것도 좋을 것 같습니다.

다만, 타입을 위한 타입(C++의 제너릭 지옥처럼)이 되지는 않도록
상황에 따른 작성자의 판단에 따라 타입 힌트의 구체성에서는 타협이 필요할 것 같습니다.
예) `list` vs `list[dict]` vs `list[dict[str, str]]`

**Reply by Kirade** [reply]
의견 감사합니다. 
타입힌트를 직접 사용해보니, 말씀해주신 이점들이 체감이 되네요.
